jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hive.jdbc;
+
+import java.math.BigDecimal;
+import java.sql.Date;
+import java.sql.Timestamp;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import org.apache.hadoop.hive.common.type.HiveIntervalDayTime;
+import org.apache.hadoop.hive.common.type.HiveIntervalYearMonth;
+import org.apache.hadoop.hive.common.type.TimestampTZUtil;
+
+import org.apache.hadoop.hive.serde2.thrift.Type;
+
+import org.apache.hive.service.rpc.thrift.TRowSet;
+import org.apache.hive.service.cli.RowSet;
+import org.apache.hive.service.cli.TableSchema;
+import org.apache.hive.service.cli.TypeDescriptor;
+
+/**
+ * ConvertedResultSet
+ */
+public class ConvertedResultSet implements RowSet {
+
+  private final ArrayList<Object[]> rows;
+  private final int numColumns;
+
+  private interface ConvertFunc {
+    Object convert(Object o);
+  }
+
+  static final int nTypes = Type.values().length;
+  static final ConvertFunc convertFuncs [] = new ConvertFunc [nTypes];
```

Comment:
why not use Map?
````
  static final Map<Type, UnaryOperator<?>> convertFuncs = new HashMap<Type, UnaryOperator<?>>() {{
      put(Type.BINARY_TYPE, o -> (o instanceof String) ? ((String) o).getBytes() : 0);
      ....
    }};
````

Reply:
Map would be less efficient than the current ordinal lookup.

Reply:
That is the usage pattern for the HashTable - do the lookups. I believe this code part does not need to be optimized at the expense of readability.
````
  static final Map<Type, UnaryOperator<Object>> convertFuncs = new HashMap<Type, UnaryOperator<Object>>() {{
    put(Type.BINARY_TYPE, o -> (o instanceof String) ? ((String) o).getBytes() : 0);
    put(Type.TIMESTAMP_TYPE, o -> Timestamp.valueOf((String) o));
    put(Type.TIMESTAMPLOCALTZ_TYPE, o -> TimestampTZUtil.parse((String) o));
    put(Type.DECIMAL_TYPE, o -> new BigDecimal((String) o));
    put(Type.DATE_TYPE, o -> Date.valueOf((String) o));
    put(Type.INTERVAL_YEAR_MONTH_TYPE, o -> HiveIntervalYearMonth.valueOf((String) o));
    put(Type.INTERVAL_DAY_TIME_TYPE, o -> HiveIntervalDayTime.valueOf((String) o));
  }};

  public ConvertedResultSet(RowSet rowSet, TableSchema schema) {
    rows = new ArrayList<>();
    numColumns = rowSet.numColumns();
    Iterator<Object[]> srcIter = rowSet.iterator();
    final int nCols = schema.getSize();
    
    List<UnaryOperator<Object>> colConvert = new ArrayList<>(nCols);
    for (int i = 0; i < nCols; ++i) {
      colConvert.add(convertFuncs.get(schema.getColumnDescriptorAt(i).getType()));
    }
    
    while (srcIter.hasNext()) {
      Object[] srcRow = srcIter.next();
      Object[] dstRow = new Object[srcRow.length];
      for (int i = 0; i < nCols; ++i) {
        if (colConvert.get(i) != null && srcRow[i] != null) {
          dstRow[i] = colConvert.get(i).apply(srcRow[i]);
        } else {
          dstRow[i] = srcRow[i];
        }
      }
      rows.add(dstRow);
    }
  }
````

Reply:
With this change there was a 1.1% decrease in performance. I can keep the change if you insist.

Reply:
@kdeschle what Map implementation have you used? EnumMap to HashMap?
EnumMap is 1-to-1 match with your solution (vals array with key.ordinal() index access)

1.1% decrease in generic performance performance or collection access?

Reply:
Nevermind. It seems it was just variation in my runs. Posted new rev with these changes and some fixes.